ic-agent/src/agent/response_authentication.rs

```diff
@@ -132,25 +137,69 @@ pub(crate) fn lookup_reply<Storage: AsRef<[u8]>>(
         request_id.as_slice(),
         "reply".as_bytes(),
     ];
-    let reply_data = lookup_value(certificate, path)?;
+    let reply_data = lookup_value(&certificate.tree, path)?;
     let arg = Vec::from(reply_data);
     Ok(RequestStatusResponse::Replied(ReplyResponse { arg }))
 }
 
+pub(crate) fn lookup_subnet<Storage: AsRef<[u8]> + Clone>(
+    certificate: &Certificate<Storage>,
+    root_key: &[u8],
+) -> Result<(Principal, Subnet), AgentError> {
+    let subnet_id = if let Some(delegation) = &certificate.delegation {
+        Principal::from_slice(delegation.subnet_id.as_ref())
+    } else {
+        Principal::self_authenticating(root_key)
+    };
+    let subnet_tree = lookup_tree(&certificate.tree, [b"subnet", subnet_id.as_slice()])?;
+    let key = lookup_value(&subnet_tree, [b"public_key".as_ref()])?.to_vec();
+    let canister_ranges = lookup_value(&subnet_tree, [b"canister_ranges".as_ref()])?;
+    let canister_ranges: Vec<(Principal, Principal)> = serde_cbor::from_slice(canister_ranges)?;
+    let node_keys_subtree = lookup_tree(&subnet_tree, [b"node".as_ref()])?;
+    let mut node_keys = HashMap::new();
+    for path in node_keys_subtree.list_paths() {
+        if path.len() < 2 {
+            // if it's absent, it's because this is the wrong subnet
+            return Err(AgentError::CertificateNotAuthorized());
+        }
+        if path[1].as_bytes() != b"public_key" {
+            continue;
+        }
+        if path.len() > 2 {
+            return Err(AgentError::LookupPathError(
+                path.into_iter()
+                    .map(|label| label.as_bytes().to_vec().into())
+                    .collect(),
+            ));
+        }
```

Comment:
We are being rather strict as to what all paths under /node look like.
If the spec changes in the future and a node can respond with other paths, this code would fail.

Should it return an error or simply skip paths which do not adhere to the /<node_id>/public_key format we expect?

Reply:
It would fail if `public_key` ever became a branch node instead of a leaf node, but there is already nothing to do about that since then it could not contain a value. And if it receives any path that is not `public_key`, it skips over it instead of failing.

Reply:
If there is ever another 1 length path under /node this would return AgentError::CertificateNotAuthorized(), right?

Anyway, not a concrete security risk at this time. Might need to be updated if the spec ever changes which might be overlooked at that point.

Reply:
Given that it goes `node/<node id>/public_key`, I do not think there is any risk of anything else going under `node`, since the whole point is to treat it as a hash-map and you could not otherwise detect whether you had a node ID or not.